### DISCUSSION TITLE
Home Assistant-Pannello energia e monitoraggio con PZEM004T-v3.0 e ESP32.

### DISCUSSION THREAD

**Discussion by Orfmia** [indomus/forum#724] [HUB: Home Assistant]
Ciao a tutti e grazie di esserci! 😊
Chiedo scusa da subito se non rispetto perfettamente le regole del forum, è la mia prima volta che scrivo, non me ne vogliate!

Con l’avvento del menù “Energia” di agosto, anch’io, entusiasta, ho provato a integrarlo nel mio Home Assistant.
Ho acquistato 4 PZEM004T-v3.0 (in cina) e ho provato a collegarli a un ESP32 già presente nel quadro elettrico (per la gestione delle luci esterne alla casa).

Ho usato la seguente libreria :  https://github.com/mandulaj/PZEM-004T-v30 per leggere attraverso ul bus seriare dell'ESP32 i  sensori, a parte un problema con il quarto PZEM che mi genera errori di lettura, tutto funziona bene!

C’ho anche fatto dei contenitori per alloggiarli nella barra DIN del quadro elettrico, (a chi interessa i file per la stampa 3d questo è il link: https://www.thingiverse.com/thing:4965538)

L’ESP32 “legge” dal Bus seriale i dati (ho copiato gli esempi della libreria “Mandulaj”) dei tre sensori funzionanti (!) e ne pubblica i dati in MQTT.

Lato Home Assistant leggo i sensori MQTT così:
`

  - platform: mqtt
    name: "Totale_Tensione"
    state_topic: "CasaMia/Energia/PZEMS0/Voltaggio"
    unit_of_measurement: 'V'
  - platform: mqtt
    name: "Totale_Frequenza"
    state_topic: "CasaMia/Energia/PZEMS0/Frequenza"
    unit_of_measurement: 'Hz'
    
  - platform: mqtt
    name: "Totale_Corrente"
    state_topic: "CasaMia/Energia/PZEMS0/Corrente"
    unit_of_measurement: 'A'
  - platform: mqtt
    name: "Totale_cosfi"
    state_topic: "CasaMia/Energia/PZEMS0/pf"
    unit_of_measurement: 'Cos φ'
  - platform: mqtt
    name: "Totale_Potenza"
    state_topic: "CasaMia/Energia/PZEMS0/Potenza"
    qos: 0
    device_class: energy
    state_class: measurement
    unit_of_measurement: 'W'
    
  - platform: mqtt
    name: "Totale_Energia"
    state_topic: "CasaMia/Energia/PZEMS0/Energia"
    qos: 0
    device_class: energy
    state_class: total_increasing
    unit_of_measurement: "kWh"

  - platform: mqtt
    name: "Solare_Tensione"
    state_topic: "CasaMia/Energia/PZEMS1/Voltaggio"
    unit_of_measurement: 'V'
  - platform: mqtt
    name: "Solare_Frequenza"
    state_topic: "CasaMia/Energia/PZEMS1/Frequenza"
    unit_of_measurement: 'Hz'
    
  - platform: mqtt
    name: "Solare_Corrente"
    state_topic: "CasaMia/Energia/PZEMS1/Corrente"
    unit_of_measurement: 'A'
  - platform: mqtt
    name: "Solare_cosfi"
    state_topic: "CasaMia/Energia/PZEMS1/pf"
    unit_of_measurement: 'Cos φ'
  - platform: mqtt
    name: "Solare_Potenza"
    state_topic: "CasaMia/Energia/PZEMS1/Potenza"
    qos: 0
    device_class: energy
    state_class: measurement
    unit_of_measurement: 'W'
  - platform: mqtt
    name: "Solare_Energia"
    state_topic: "CasaMia/Energia/PZEMS1/Energia"
    qos: 0
    device_class: energy
    state_class: total_increasing
    unit_of_measurement: "kWh"
`

Tutto perfetto!
Nel menù Energia trovo, e posso inserire, i sensori (le entità MQTT) caratterizzati con state_class: total_increasing:
**Totale_Energia
Solare_Energia**

Ora la mia perplessità:
Ho posizionato due sensori uno all’ingresso del quadro elettrico di casa che misura tutta l’energia che “consumo” e uno all’uscita dell’inverter che misura la produzione solare (a monte / verso rete del precedente).

ENEL
 |
 | 
O------------> **Totale Energia**    ---> Impianto di Casa
 |
 ---------- <--    **Solare Energia** <---  Pannelli Solari

Non ho messo un sensore all’uscita del contatore ENEL perché i sensori PZEM004T rilevano molti dati ma non il “verso” della corrente, (ovvio forse, è corrente alternata!), questo quindi non mi permette di capire se il verso della corrente è dalla rete verso casa mia (consumo pagato all’Enel) o se sto invece fornendo io (con i pannelli solari) energia alla rete.

Volendo posso creare un sensore “virtuale” che mi restituisca la differenza algebrica tra l’energia prodotta dai pannelli e quella consumata che potrebbe darmi l’energia restituita alla rete:
Return_to_Gred = **Solare _Energia – Totale_Energia**

Ma Home Assistant lavora sul “totale dell’energia” (total_increasing) e non sul dato in tempo reale e non capisco come poter integrare in Home Assistant queste misure e quale è la logica di analisi e di calcolo 

In altre parole se i pannelli solari producono 1000 e a casa consumo 200 il meccanismo potrebbe anche funzionare

Return_to_Gred = **Solare _Energia – Totale_Energia**
Return_to_Gred = 1000 – 200 = 800 
Return_to_Gred =800 😊

Ma se i pannelli solari non producono niente (di notte) e mia moglie fa la torta con il forno elettrico:

Return_to_Gred = **Solare _Energia – Totale_Energia**
Return_to_Gred = 0 – 2500 = -2500 
Return_to_Gred = -2500 ☹

che succede di quel dato negativo, trattandosi di total_increasing mi decrementa il consumo totale??
Qualcuno ha le idee più chiare di me e riesce a spiegarmi come posso fare per avere una analisi corretta anche del "Return to Greed?
Grazie!
Orfmia

**Comment by Orfmia** [comment]
Mi rispondo da solo :-)
 ho calcolato il valore dell'energia del "Ritorno alla rete" come differenza tra prodotto dai pannelli solari e consumato dalla casa.
con questo dato Home Assistant funziona ma fa confusione nel gestire i valori "negativi" relativi all'immissione in rete dai pannelli solari: di notte, l'inverter assorbe circa mezzo Watt e questo fa decrementare il valore dell'energia immessa in rete 
per risolvere il problema ho inserito nel codice dell'ESP un controllo che impedisce questo "decremento" dell'energia immessa in rete e così facendo funziona tutto regolarmente:

![Capture](https://user-images.githubusercontent.com/42579888/134953659-775830bb-bd88-4b28-816e-e60770767388.JPG)


Nota: i valori degli istogrammi qui sopra evidenziano che in un particolare orario c'è stato sia "prelievo dalla rete" in blu-grigio che immissione in rete: Viola.
Non si tratta di una contraddizione ma della media del periodo, che può vedere momenti di assorbimento tale da superare le capacità produttive dei pannelli (in giallo nel grafico sotto) e poi, qualche minuto dopo l'immissione in rete perché il fabbisogno di energia è inferiore a quanto prodotto dal Solare.

**Comment by Giampy72** [comment]
ciao..
ho anche io un problema simile
in realtà uso due shelly em che prima dell'aggiornamento di HA mi funzionavano bene
ora non vedo piu i dati , nonostante nel pannello energia riesco ad associare gli shelly
probabilmente occorre cambiare la configurazione dei sensori ma non riesco a risolvere.
li uso senza mqtt in questo modo

  customize:
    sensor.shelly_shem_98cdac1f28eb_1_total_consumption:
      last_reset: "2022-03-11T00:00:00+00:00"
      device_class: energy
      state_class: measurement
      unit_of_measurement: "kWh"

    sensor.shelly_shem_98cdac38f892_1_total_consumption:
      last_reset: "2022-03-11T00:00:00+00:00"
      device_class: energy
      state_class: measurement
      unit_of_measurement: "kWh"

il primo misura la corrente consumata
mentre il secondo misura quella prodotta dai pannelli

dopo l'aggiornamento non vedo piu i dati nel pannello energia
